Migration of  "Design > Pages" page cms category listing

tomas862 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migration of  "Design > Pages" CMS pages. There are 2 listings in this page, this PR  **only migrates the top listing: categories** and the **breadcrumbs** related with it above. Page can be reached via **admin-dev/index.php/improve/design/cms-pages**
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10565
| How to test?  | See below

## How to test

There are 2 listings in this page. This PR migrates the top listing: categories. It also migrates thebreadcrumbs.
The **CMS category list** and breadcrumbs should work exactly the same as in legacy page (excluding position drag & drop action).
Please note that
- everything regarding the second listing
- positions drag and drop and search by position

will be migrated in another PR.
Access migrated page through `admin-dev/index.php/improve/design/cms-pages`

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/10877)
<!-- Reviewable:end -->

Commit:
adds empty controller, adds default cmsCategoryId in url

Commit:
displays cms categories list with default data

Commit:
renames cms to cmspages, adds required wildcard

tomas862 committed:
fixes sorting

tomas862 committed:
adds filters definiton

Commit:
adds edit page route

Commit:
generates route with cms parent category id

Commit:
uses event listener to modify filters

Commit:
implements filters functionality

Commit:
adds extra route param fields which can be passed to generate urls with more then one variable

Commit:
adds delete route

Commit:
renamed cms pages to cms page

Commit:
adds cms page category deletion command

Commit:
implements usage of delete cms page command

Commit:
creates cms page category status toggling command

Commit:
implements toggle cms page category status usage

Commit:
updates positions so it has available routeParamNames to set

Commit:
adds some todos

Commit:
renames some routes, adds bulk delete cms page category command handler

Commit:
implements bulk delete command

Commit:
adds bulk enable command and its implementation

Commit:
implements disable status command usag

PierreRambaud commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkEnableCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,70 @@
+        try {
+            foreach ($command->getCmsPageCategoryIds() as $cmsPageCategoryId) {
+                $entity = new CMSCategory($cmsPageCategoryId->getValue());
+                $entity->active = 1;
```
use true instead of 1, it's a boolean =)

Commit:
adds embeded footer

Commit:
implements back to root cms page category list action

Commit:
create new page button, adds grid actions

Commit:
adds breadcrumb query bus handler

tomas862 committed:
implements breadcrumbs usage

tomas862 committed:
removes unused parameters

tomas862 committed:
adds view data provider

tomas862 committed:
added permissions checking

tomas862 committed:
added missing comments, renamed some classes, added some todos

tomas862 commented:
waiting for position-extension javascript to be available on development branch and some minor improvements for DemoRestricted and AminSecurity annotations I hope to introduce in another PR

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkDeleteCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,72 @@
+                $entity = new CMSCategory($cmsPageCategoryId->getValue());
+
+                if (false === $entity->delete()) {
+                    throw new CannotBulkDeleteCmsPageCategoryException(
```
does it make sense to have separate exception for bulk? I dont think it does and here's why:

1. when it is thrown i expect it to have multiple CMS page category ids, since it is used for bulk action.
2. if it failed to delete cms category id, cant you just throw `CannotDeleteCmsPageCategoryException`, you dont care about context (whether its bulk, single or any other), do you?

same applies to other handlers.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkDisableCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Class BulkDisableCmsPageCategoryHandler is responsible for deleting multiple cms page categories.
+ */
+class BulkDisableCmsPageCategoryHandler implements BulkDisableCmsPageCategoryHandlerInterface
```
`final`

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkDisableCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,70 @@
+        try {
+            foreach ($command->getCmsPageCategoryIds() as $cmsPageCategoryId) {
+                $entity = new CMSCategory($cmsPageCategoryId->getValue());
+                $entity->active = false;
```
you should throw exception if `CMSCategory` is not found before trying to update it.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkEnableCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,70 @@
+        try {
+            foreach ($command->getCmsPageCategoryIds() as $cmsPageCategoryId) {
+                $entity = new CMSCategory($cmsPageCategoryId->getValue());
+                $entity->active = true;
```
same here, make sure `CMSCategory` exists.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/DeleteCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,91 @@
+    public function handle(DeleteCmsPageCategoryCommand $command)
+    {
+        try {
+            $entity = new CMSCategory($command->getCmsPageCategoryId()->getValue());
```
you have 2 try/catch statements in this methods, can you wrap them into single?

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/ToggleCmsPageCategoryStatusHandler.php:
```diff
@@ -0,0 +1,91 @@
+     */
+    public function handle(ToggleCmsPageCategoryStatusCommand $command)
+    {
+        try {
```
same here, why 2 try/catch statements?

sarjon commented on src/Adapter/CMS/PageCategory/QueryHandler/GetCmsPageCategoriesForBreadcrumbHandler.php:
```diff
@@ -0,0 +1,98 @@
+
+use CMSCategory;
+use Doctrine\DBAL\Connection;
+use function ICanBoogie\array_insert;
```
i think this should not be included here. :smile:

sarjon commented on src/Adapter/CMS/PageCategory/QueryHandler/GetCmsPageCategoriesForBreadcrumbHandler.php:
```diff
@@ -0,0 +1,98 @@
+     */
+    private $contextLanguageId;
+
+    public function __construct($contextLanguageId)
```
docblock missing

sarjon commented on src/Core/Domain/CmsPageCategory/Query/GetCmsPageCategoriesForBreadcrumb.php:
```diff
@@ -0,0 +1,59 @@
+ * Class GetCmsPageCategoriesForBreadcrumb is responsible for providing required data for displaying cms page category
+ * breadcrumbs.
+ */
+class GetCmsPageCategoriesForBreadcrumb
```
i dont really feel good about this query. :/ @matks wdyt, is it a valid use case for this query?

matks replied:
@tomas862 The idea behind Queries is rather to gather all the data to display 1 page. A Query is a usecase. This looks like a part of a page 🤔

tomas862 replied:
@matks do you mean to have a query named like "GetDataForCmsPage" which returns everything (except lists) on that page? :thinking:  
The thing is that its only one custom small part on that page that needs to be returned  - the **breadcrumbs**. I saw the customer page sample where a lot of data is being returned - in that case I would agree but for such case I don't know if its requried to do so :thinking:

matks replied:
My bad 👍 I did not see the context. You're right 😄 this query is valid

sarjon commented on src/Core/CMS/CmsPageViewDataProvider.php:
```diff
@@ -0,0 +1,88 @@
+    /**
+     * @param CommandBusInterface $queryBuss
+     */
+    public function __construct(CommandBusInterface $queryBuss)
```
```diff
- $queryBuss
+ $queryBus
```

sarjon commented on src/Core/CMS/CmsPageViewDataProvider.php:
```diff
@@ -0,0 +1,88 @@
+/**
+ * Class CmsPageViewDataProvider provides cms page view data for cms listing page.
+ */
+class CmsPageViewDataProvider
```
you _must_ define interface for it and make it final.

sarjon commented on src/Core/CMS/CmsPageViewDataProvider.php:
```diff
@@ -0,0 +1,88 @@
+    public function getView($cmsCategoryParentId)
+    {
+        return [
+            'rootCategoryId' => CmsPageRootCategorySettings::ROOT_CMS_PAGE_CATEGORY_ID,
```
```diff
- rootCategoryId
+ root_category_id
```

sarjon commented on src/Core/CMS/CmsPageViewDataProvider.php:
```diff
@@ -0,0 +1,88 @@
+        try {
+            $cmsPageCategoryId = new CmsPageCategoryId($cmsCategoryParentId);
+            /** @var CmsPageCategoriesBreadcrumbTree $result */
+            $result = $this->queryBuss->handle(new GetCmsPageCategoriesForBreadcrumb($cmsPageCategoryId));
```
`$result` is not really the best name here.

sarjon commented on src/Core/CMS/CmsPageViewDataProvider.php:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @param int $cmsCategoryParentId
+     *
+     * @return array
```
i prefer Dto to be returned instead, @matks wdyt?

sarjon commented on src/Core/Domain/CmsPageCategory/CmsPageCategoriesBreadcrumbTree.php:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Class CmsPageCategoriesForBreadcrumb is a DTO which is used in building cms page categories breadcrumb.
+ */
+class CmsPageCategoriesBreadcrumbTree
```
i dont really see much value in this Dto as you can put any array in it and then retrieve it, @matks wdyt?

sarjon commented on src/Core/Domain/CmsPageCategory/Exception/CannotBulkDeleteCmsPageCategoryException.php:
```diff
@@ -0,0 +1,64 @@
+
+namespace PrestaShop\PrestaShop\Core\Domain\CmsPageCategory\Exception;
+
+use Throwable;
```
watch out, its php 7+.

sarjon commented on src/Core/Domain/CmsPageCategory/Exception/CannotBulkDeleteCmsPageCategoryException.php:
```diff
@@ -0,0 +1,64 @@
+        $message = "",
+        $code = 0,
+        Throwable $previous = null,
+        $cmsPageCategoryId
```
it is **really bad practice** to put mandatory parameters after optional, i suggest you put it as a first parameter. in addition to that, how about accepting value object instead of int?

sarjon commented on src/Core/Domain/CmsPageCategory/Exception/CannotBulkDeleteCmsPageCategoryException.php:
```diff
@@ -0,0 +1,64 @@
+    public function __construct(
+        $message = "",
+        $code = 0,
+        Throwable $previous = null,
```
same here with `Throwable`

sarjon commented on src/Core/Domain/CmsPageCategory/ValueObject/CmsPageCategoryId.php:
```diff
@@ -0,0 +1,60 @@
+        $this->cmsPageCategoryId = (int) $cmsPageCategoryId;
+    }
+
+    public function getValue()
```
docblock is missing

sarjon commented on src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php:
```diff
@@ -0,0 +1,317 @@
+        UrlGeneratorInterface $urlGenerator,
+        RequestStack $requestStack
+    ) {
+        $this->resetActionUrl =  $urlGenerator->generate('admin_common_reset_search', [
```
i prefer to generate `$resetActionUrl` when needed instead of saving state on service, @matks wdyt?

sarjon commented on src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php:
```diff
@@ -0,0 +1,317 @@
+            $this->cmsCategoryParentId = $request->attributes->get('cmsCategoryParentId');
+        }
+
+        $this->redirectionUrl = $urlGenerator->generate('admin_cms_pages_index', [
```
same here.

sarjon commented on src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php:
```diff
@@ -0,0 +1,317 @@
+        ;
+    }
+
+    public function getBulkActions()
```
docblock is missing

sarjon commented on src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php:
```diff
@@ -0,0 +1,317 @@
+        ;
+    }
+
+    public function getGridActions()
```
docblock is missing.

sarjon commented on src/Core/Grid/Filter/CmsPageCategoryFilterFormFactory.php:
```diff
@@ -0,0 +1,103 @@
+/**
+ * Class CmsPageCategoryFilterFormFactory is responsible for changing form action to the custom one.
+ */
+class CmsPageCategoryFilterFormFactory implements GridFilterFormFactoryInterface
```
should be `final`

sarjon commented on src/Core/Grid/Filter/CmsPageCategoryFilterFormFactory.php:
```diff
@@ -0,0 +1,103 @@
+            );
+        }
+
+        $request = $this->requestStack->getCurrentRequest();
```
maybe its better to rely on master request as subrequests may not contain the same data? but im really not sure here.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+            $filters = $searchParametersForm->getData();
+        }
+
+        return $this->redirectToRoute('admin_cms_pages_index', compact('cmsCategoryParentId', 'filters'));
```
even though `compact` seems like a cool thing to use, i would not recommend doing so, mainly because if you were to change `$cmsCategoryParentId` variable name and forget to update `compact` params, you wont get any error as it silently skips parameters that are not defined, so it may hide bugs that are really hard to track down.

second, it forces you to make array keys `camelCase` which i think is against Symfony's coding standard (but i may be wrong regarding that).

@matks wdyt about it? personally, i dont want to encourage use of `compact` in core. :/

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+     * Deletes multiple cms page categories.
+     *
+     * @AdminSecurity("is_granted('delete', request.get('_legacy_controller'))", message="You do not have permission to delete this.")
+     *
```
dont forget `DemoRestricted` access control.

tomas862 replied:
want to introduce one change regarding that in another PR due to DemoRestricted annotation is not very suitable in my case for now

matks replied:
@tomas862 you need to tell us more 😄

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+     * Deletes cms page category and all its children categories.
+     * //todo: modify annotations so they have current controller routes
+     * @AdminSecurity("is_granted('delete', request.get('_legacy_controller'))", message="You do not have permission to delete this.")
+     *
```
dont forget `DemoRestricted` access control.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+     * Toggles cms page category status.
+     *
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", message="You do not have permission to edit this.")
+     *
```
dont forget `DemoRestricted` access control.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+        $cmsCategoriesToEnable = $request->request->get('cms_page_category_bulk');
+
+        try {
+            $this->getCommandBus()->handle(new BulkEnableCmsPageCategoryCommand($cmsCategoriesToEnable));
```
dont you need any success message?

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+     * Changes multiple cms page category statuses to enabled.
+     *
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", message="You do not have permission to edit this.")
+     *
```
same here regarding Demo access, check other methods as well.

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+        $cmsCategoriesToDisable = $request->request->get('cms_page_category_bulk');
+
+        try {
+            $this->getCommandBus()->handle(new BulkDisableCmsPageCategoryCommand($cmsCategoriesToDisable));
```
same here, no success message?

tomas862 replied:
legacy controller did not had any success message but maybe you're right here - seems to look better with message :+1:

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+     *
+     * @return string
+     */
+    private function handleException(DomainException $exception)
```
you only catch `CmsPageCategoryException`, so why accepting `DomainException` as parameter?

tomas862 replied:
My controller here will have two lists so later on there will be more different exceptions to handle

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+        $errorMessage = $this->trans('Unexpected error occurred.', 'Admin.Notifications.Error');
+        $statusCode = $exception->getCode();
+
+        if ($exception instanceof CmsPageCategoryException && 0 === $statusCode) {
```
based on your code, its always instance of `CmsPageCategoryException`

sarjon commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,378 @@
+     *
+     * @return string
+     */
+    private function getCmsPageCategoryErrorByExceptionTypeAndCode(CmsPageCategoryException $exception, $statusCode)
```
why do you pass code if `$exception` itself contains code?

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/improve/design/cms_pages.yml:
```diff
@@ -0,0 +1,77 @@
+        _controller: 'PrestaShopBundle:Admin\Improve\Design\CmsPage:searchCategory'
+        _legacy_controller: AdminCmsContent
+
+admin_cms_pages_create_cms_category:
```
you can make it shorter:

```diff
- admin_cms_pages_create_cms_category
+ admin_cms_categories_create
```

same for other routes.

tomas862 replied:
does it not break the naming convention ? My route is /cms-pages so all names should start with admin_cms_pages etc... ?

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/improve/design/cms_pages.yml:
```diff
@@ -0,0 +1,77 @@
+        _legacy_controller: AdminCmsContent
+
+admin_cms_pages_create_cms_category:
+    path: /category/create
```
```diff
- /category/create
+ /categories/new
```

make sure you use plural for naming resources.

sarjon commented on src/PrestaShopBundle/Resources/config/routing/admin/improve/design/cms_pages.yml:
```diff
@@ -0,0 +1,77 @@
+
+admin_cms_pages_delete_cms_category:
+    path: /category/{cmsCategoryId}/delete
+    methods: DELETE
```
i prefer to be consistent and use `[DELETE]` even though it accepts only 1 method, but its personal preference so i cant force that. :)

p.s. also notice method**s**, its in plural, so maybe array makes sense. :p

sarjon commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/index.html.twig:
```diff
@@ -0,0 +1,63 @@
+  *#}
+
+{% extends '@PrestaShop/Admin/layout.html.twig' %}
+{% trans_default_domain "Admin.Design.Feature" %}
```
remember, this may not work for translations.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkDeleteCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,72 @@
+    {
+        try {
+            foreach ($command->getCmsPageCategoryIds() as $cmsPageCategoryId) {
+                $entity = new CMSCategory($cmsPageCategoryId->getValue());
```
check if it exists before deleting.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkDeleteCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,72 @@
+                        ),
+                        0,
+                        null,
+                        $cmsPageCategoryId->getValue()
```
please move it to first argument, see comment below regarding arguments order.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/BulkDisableCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,70 @@
+                $entity->active = false;
+
+                if (false === $entity->update()) {
+                    throw new CannotBulkDisableCmsPageCategoryException(
```
it would be nice to set `id` of cms page category on exception, which failed to disable. same for other commands.

sarjon commented on src/Adapter/CMS/PageCategory/CommandHandler/DeleteCmsPageCategoryHandler.php:
```diff
@@ -0,0 +1,91 @@
+        }
+
+        if (0 >= $entity->id) {
+            throw new CmsPageCategoryNotFoundException(
```
it would be nice to know on exception which page category it didnt find.

tomas862 replied:
don't you think if I gonna pass the id via the constructor I should use it in the controller? In case of bulk disable/enable , deleting single does not require to have an id. And anyway, the exception message has an id so I don't think it's required in all cases :thinking:

sarjon reviewed (commented):
huuh, quite big PR. :)

regarding Domain name, is it the right one? i see `CMS` as domain which has `Category` and `Role` as subdomains. inside `Category` there is another `Page` subdomain, @matks need your input.

Commit:
removes non-required exceptions

Commit:
reduces try catches count

Commit:
uses value object for building iteratable result for breadcrumbs

Commit:
used object values instead of plain id

Commit:
improves grid definition by calling resetActionUrl and redirectionUrl generations where required

tomas862 committed:
adds class as final, removes compact php feature usage and replaces it with regular arrays

tomas862 committed:
changes some naming

tomas862 commented:
Thanks @sarjon for the review!

matks commented:
@tomas862 https://github.com/PrestaShop/PrestaShop/pull/11075 has been merged into `develop` :)

Commit:
adds improved AdminSecurity and DemoRestricted annotations usage with query parameters to keep

tomas862 commented:
Waiting for PrestaShop 1.7.5 to be merge into develop due to missing positions javascript

matks commented:
@tomas862 1.7.5.x has been merged into develop :) https://github.com/PrestaShop/PrestaShop/pull/11072

Commit:
fixes position column usage

Commit:
improves link edit by changing route params generation to get parent id information from database

Commit:
improves toggle column by changing route params generation to get parent id information from database

Commit:
improves edit link generation  by changing route params generation to get parent id information from database

Commit:
improves submit route actiob  generation  by changing route params generation to get parent id information from database

Commit:
adds cms page category position update implementation

Commit:
adds admin Security and demo restricted annotations

Commit:
adds admin Security and demo restricted annotations

Commit:
adds missing comment

Commit:
removes non-required status code

tomas862 commented:
Hi @matks 

I have a question related with current Pull reuqest:
If I filter by position, for instanse I enter **1** it filters the record which has the position **2** - same happens on legacy list due to the positions in database starts from 0 and not from 1 :smile:  Should I think of solutions of how to fix it or is it okey with that?

matks commented:
Hi @tomas862 sorry for the late answer
We need to fix this bug 😄 but in another PR in order to keep Pull Requests manageable.
So let's finish and merge this PR *then* fix the bug.

We will not forget because it is listed in the Issue summary: https://github.com/PrestaShop/PrestaShop/issues/10565#issue-362232806

tomas862 committed:
changes some position column parameters, renames id of list so its now lowercase

tomas862 commented:
this Pull request is ready for the review:
only the first list which includes cms categories is implemented with all actions (except position drag & drop functionality which will be improved with another PR and some bugs that are mentioned under current issue)
Also breadcrumbs are included

assets needs to be recompiled in order to test the functionality

mickaelandrieu commented:
Hello @tomas862,

would you mind to fix the lint errors?

Just execute the following command:

```
./vendor/bin/php-cs-fixer fix
```

Mickaël

tomas862 committed:
adds todos

tomas862 committed:
applies cs php fixer

tomas862 commented:
ready for review - rebuilding of the assets is required to check the functionality :slightly_smiling_face:

tomas862 committed:
improves search and reset type routes assignation

matks commented on src/Core/Domain/CmsPageCategory/ValueObject/CmsPageCategory.php:
```diff
@@ -0,0 +1,69 @@
+namespace PrestaShop\PrestaShop\Core\Domain\CmsPageCategory\ValueObject;
+
+/**
+ * Class CmsPageCategory is responsible for providing data in a comfortable way.
```
Comfort is important indeed 😄

matks commented on src/Core/Domain/CmsPageCategory/Command/ToggleCmsPageCategoryStatusCommand.php:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * @param CmsPageCategoryId $cmsPageCategoryId
+     */
+    public function __construct(CmsPageCategoryId $cmsPageCategoryId)
```
After discussing it with eternoendless, we would like all Commands to only accept primitive php types as input. The idea is to make them as simple as possible to use.

One way to do it is to accept integers as input, but convert them into the right type so you benefit from ValueObject properties 😉 
```
public function __construct($cmsPageCategoryId)
    {
    	if (false === ($cmsPageCategoryId instanceof CmsPageCategoryId::class)) {
    		$cmsPageCategoryId = new cmsPageCategoryId($cmsPageCategoryId);
    	}

        $this->cmsPageCategoryId = $cmsPageCategoryId;
    }
```

matks commented on src/Core/Domain/CmsPageCategory/Exception/CannotDeleteCmsPageCategoryException.php:
```diff
@@ -0,0 +1,61 @@
+     * @param int $code
+     * @param null $previous
+     */
+    public function __construct(CmsPageCategoryId $cmsPageCategoryId, $message = '', $code = 0, $previous = null)
```
Same for exceptions: let's make these new classes as simple as possible to use for prestashop developers ;) by using primitive php types as input, but using advanced objects internally

matks commented on src/Core/Domain/CmsPageCategory/ValueObject/CmsPageCategoryId.php:
```diff
@@ -0,0 +1,63 @@
+                sprintf('Invalid cms page category id %s supplied', var_export($cmsPageCategoryId, true))
+            );
+        }
+        $this->cmsPageCategoryId = (int) $cmsPageCategoryId;
```
I'm wondering ... maybe we should just refuse not-integer values for `$cmsPageCategoryId` ? Forcing people to provide a real integer will make their code cleaner. Accepting values such as `"42"` looks kind but expose them to later errors

tomas862 replied:
I can agree with you - just there will be more type casting when the id will be passed for now since the ObjectModel instead of giving integer it gives string if I recall correctly

matks replied:
If ObjectModel provides bad input, it should be treated as bad and processed accordingly 😄

matks commented on src/Core/Grid/Action/Bulk/Type/SubmitBulkAction.php:
```diff
@@ -54,9 +54,12 @@ protected function configureOptions(OptionsResolver $resolver)
             ->setDefaults([
                 'confirm_message' => null,
                 'submit_method' => 'POST',
+                'route_params_extra' => [],
```
I understand your need to pass extra params however I'm a bit worried whenever in a software we start adding "extra" things 😅 . Most of the time - not always - it's a shortcut solution, not the best one.

@mickaelandrieu @sarjon an opinion ? 😄

tomas862 replied:
yeah - I used it a lot since for my case , its mandatory to pass an extra id to all routes I am generating - I'm thinking now that I can just rename it to `record_route_params` and in my scenario, instead of passing integer value I can pass associate column which is called `id_parent`.

The sample of such column would be `PositionColumn` https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Column/Type/Catalog/PositionColumn.php#L65

what do you think about that ? :)

tomas862 replied:
resolved - `route_params_extra` is no longer in use

matks commented on src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php:
```diff
@@ -0,0 +1,319 @@
+     */
+    public function __construct(RequestStack $requestStack)
+    {
+        $this->setCmsPageCategoryParentId($requestStack);
```
Note for myself: find a way to get rid of RequestStack in SF services

I know you need it here so let's not change the code, but we should really find a way not to rely on request

matks commented on src/Core/Grid/Filter/CmsPageCategoryFilterFormFactory.php:
```diff
@@ -0,0 +1,103 @@
+        $request = $this->requestStack->getCurrentRequest();
+
+        if (null !== $request) {
+            $newActionUrl = $this->urlGenerator->generate('admin_cms_pages_search_cms_category', [
```
So this is how you replace "extra route parameters" with their real value ?

The idea is interesting. I'm just worried that it's not obvious for someone that passing `route_params_extra` parameters is not enough, and that a 2nd class (CmsPageCategoryFilterFormFactory here) is needed to perform this extra processing.

1) Do you think this behavior could be used in other places ? If yes we might define a specific global behavior such as "if you pass parameters with name request_param_cmsCategoryParentId, it will be automatically casted from the request bag"

2) Alternative: if the behavior is not global and requires to use a specific FormFactory, if the FormFactory is not declared we throw an Exception. So we oblige developers to add it.

3) If we dont do 1) or 2), we need at least to explain what's going to happen in the Grid Definition with code comments in the Definition and a php class doc more verbose here: https://github.com/PrestaShop/PrestaShop/pull/10877/files#diff-6d292ee7cbe1fde98b5b802ba4883b7fR36

tomas862 replied:
@matks can you explain point 2 in more details  :thinking: 

This whole class is needed just to override grids form and add action attribute to the form in order to keep the id in url  after the form submit: admin-dev/index.php/improve/design/cms-pages/1

This functionality will be needed in all lists which have nested lists inside and want to pass ids like in: admin-dev/index.php/improve/design/cms-pages/1.

If we want to make this form easily reusable maybe we should put the usage in GridDefinition classes ?

Right now  `CmsPageCategoryFilterFormFactory.php` is a [Decorator](https://sourcemaking.com/design_patterns/decorator/php) for form default factory.
@sarjon do you have any suggestions to keep it how it is or introduce something new?

matks replied:
For this PR let's keep it like this. Next nested list we handle we will see how to make it generic, with https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Filter/CategoryFilterFormFactory.php too

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,501 @@
+     *
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @Template("@PrestaShop/Admin/Improve/Design/Cms/index.html.twig")
```
Please use `$this->render()` 😉

matks commented on src/PrestaShopBundle/EventListener/FilterCmsPageCategorySearchCriteriaListener.php:
```diff
@@ -0,0 +1,82 @@
+ * Class FilterCmsPageCategorySearchCriteriaListener is responsible for updating CmsCategoryFilters filter with
+ * cms page category id.
+ */
+class FilterCmsPageCategorySearchCriteriaListener
```
I might be missing something but why not performing this in the Controller action https://github.com/PrestaShop/PrestaShop/pull/10877/files#diff-dee4848ddf509adc016fd3e905b03f37R91 ?

Splitting the behavior in 2 files make it harder to understand ?

tomas862 replied:
Well I might agree with you here - using the event listener is cool way to change default results in this case but it might be unfamiliar for most people where that modified content comes from .
@sarjon  what do you think?

PierreRambaud commented on admin-dev/themes/new-theme/js/pages/cms-page/index.js:
```diff
@@ -0,0 +1,53 @@
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+import Grid from "../../components/grid/grid";
```
Single quote pleaseee :)

tomas862 committed:
instead of passing an object via constructor passes scalar value

tomas862 committed:
adds common abstraction for bulk commands, casts arrays to integer arrays before using commands

tomas862 committed:
passes scalar value as well to the delete cms page exception

tomas862 committed:
replaces double quotes to single quotes

tomas862 committed:
after rebase moves cms page category list service definitions to separate files and create new position definition yml file

tomas862 committed:
on not found id it redirects to the default cms page, in another cases it hides exceptions

tomas862 committed:
applies php cs fixer

tomas862 committed:
removes mandatory id cms page from url parameter, renders cms page list data

tomas862 committed:
displays inner page lists with query url argument instead of the forced one

tomas862 committed:
fixes search action

tomas862 commented on src/Core/Grid/Definition/Factory/CmsPageCategoryDefinitionFactory.php:
```diff
@@ -0,0 +1,318 @@
+                    'route_param_name' => 'cmsCategoryId',
+                    'route_param_field' => 'id_cms_category',
+                    'route_params_extra' => [
+                        'id_parent' => 'cmsCategoryParentId',
```
@matks If I'm using query parameter such as `id_cms_category` when generating lists I want to keep thay query parameter no matter what if it exists. If customer goes to edit page of the cms category then, instead of just having `id_cms_category` I need to be aware also of the parent `id_cms_category` so the client after form submit can return to the parent list and not to the very top list.

The `LinkColumn` was very limited to just passing singe route parameter.
Do you think `route_params_extra` approach is good? Maybe i only need to rename it to `route_params` or something?

matks replied:
You mean the following scenario ?
- you are on grid, showing CMS pages & categories of Category "Clothes" which is a subcategory of "Home"
- you click on "edit" for category "Shirts" which is a subcategory of "Clothes"
- you edit this category "Shirts"
- when you submit you go back to grid, showing CMS pages & categories of Category "Clothes"

Is that right ?

I checked the legacy behavior: currently it uses the CMS Category ID of the page being edited. So on last step, instead of showing the "Clothes" grid, the legacy code shows the "Shirts" category.
Changing this behavior can be a good idea but we need Product approval and it deserves its own issue / PR 😄 . Can you create one and ping colinegin for it ?

tomas862 replied:
We have an issue with Add/Edit cms page category. I guess when someone will raise the PR which is related with that issue then  the product team can decide what to do. Right now I have one clean solution in mind which does not require to introduce any `route_params_extra` stuff and can be easily changed if product team will desire to change where page should be redirected after the add/edit action  :+1:

tomas862 committed:
fixes reset button redirect

tomas862 committed:
removes bulk action extra route params, passes parent id from bulk action delete which helps to stay on the same page after the redirect

tomas862 committed:
fixes bulk enable routes redirect action

tomas862 committed:
adds bulk disable redirect fix

tomas862 committed:
removes routes param extra from submit row action, fixes route generation on delete action

tomas862 committed:
passes different query parameter to cms page position update

tomas862 committed:
removes route params usage in redirect actions to edit page

tomas862 committed:
removes route params extra usage, fixes toggle status update redirect action

tomas862 committed:
fixes back to list button redirect

tomas862 committed:
adds shared function which redirects to cms page category index page

tomas862 committed:
hides back button on root cms page category

tomas862 committed:
fixes breadcrumb generation

tomas862 committed:
adds reset url generation fix which now does not include id_cms_category as 0 value

tomas862 committed:
fixes code styles

tomas862 commented:
@matks introduced some small changes - it is ready for re-review ;)

matks commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -0,0 +1,510 @@
+        $cmsCategoryParentId = null;
+        try {
+            /** @var CmsPageCategoryId $cmsCategoryParentId */
+            $cmsCategoryParentId = $this->getCommandBus()->handle(
```
I can see why you did it but ... no 😅 . A Command that deletes one category is not supposed to return the parent ID. This does not make sense in the domain. The domain should not solve the app layer (controllers, templating, routing ...) issues.

You can:
- retrieve the parent ID before
or
- return user on root listing

tomas862 replied:
yeah, I got your point. :laughing: 
I think I should make another Query command which would return parent id by child id, wdyt?

matks commented:
@tomas862 1 issue to solve 😉 
https://github.com/PrestaShop/PrestaShop/pull/10877#pullrequestreview-195619469

tomas862 committed:
uses separate command to retrieve cms page category id for redirection after each action

tomas862 committed:
adds placeholders for the inputs

tomas862 committed:
applies php cs fixer

tomas862 commented:
@matks  done :)

matks reviewed (dismissed):
Thanks @tomas862 😄 going for QA now

mbadrani commented:
the correct url to access to the page is: admin-dev/index.php/improve/design/cms-pages because the one mentioned in the description redirect to the dashboard page

mbadrani commented:
weird behaviour https://drive.google.com/file/d/1OI6jagtZDbNE3wJiUlB0EbWBmDBP_5X2/view

tomas862 committed:
adds help sidebar appearance

tomas862 committed:
adds cms page entry after rebase

tomas862 commented:
@mbadrani fixed the problem you have mentioned above. 
After rebase noticed that all lists have such appearance when action dots are clicked:
![out_of_bounds](https://user-images.githubusercontent.com/17051250/52006342-8d12d880-24d4-11e9-8ddf-1fa686dadebb.png)
Since it exists for all lists someone or me will fix that in separate Pull request :+1:

tomas862 committed:
adds legacy controller name from route name

marionf commented:
@tomas862 

1) Drop down actions are truncated

![capture d ecran_955](https://user-images.githubusercontent.com/13449658/52128263-c19da600-2634-11e9-8212-fc638c634b2d.png)

![capture d ecran_954](https://user-images.githubusercontent.com/13449658/52128264-c2363c80-2634-11e9-8c84-802f928f2c33.png)


2) @TristanLDD this link is currently the edit action, before on legacy page when I clicked on the line it was the view action. Should we change it ?

![capture d ecran_956](https://user-images.githubusercontent.com/13449658/52128462-50aabe00-2635-11e9-8b21-019e1c185b28.png)


@TristanLDD can you verify if this page is ok for you ?

![capture d ecran_957](https://user-images.githubusercontent.com/13449658/52128418-35d84980-2635-11e9-8eda-c5b665fcfb49.png)

matks commented:
@marionf I think it's related to https://github.com/PrestaShop/PrestaShop/issues/12333

TristanLDD commented:
Hello @marionf 

You're right she should replace the link by the view action as it is the main action of the line

Also, in legacy page we had a edit button next to the main category which doesn't appear anymore after migration: 
![capture d ecran 2019-02-05 a 12 04 48](https://user-images.githubusercontent.com/27484094/52269412-455dd800-293e-11e9-87ee-1f65e576356d.png)

Can we integrate that button like that ?
![capture d ecran 2019-02-05 a 12 14 33](https://user-images.githubusercontent.com/27484094/52269930-b9e54680-293f-11e9-9828-140639215311.png)

Thanks !

tomas862 committed:
adds cms page rdirect to view action instead of edit page form

tomas862 commented:
@marionf, @TristanLDD  regarding the comments in https://github.com/PrestaShop/PrestaShop/pull/10877#issuecomment-459738302 :

2. Done

regarding the comment https://github.com/PrestaShop/PrestaShop/pull/10877#issuecomment-460600871:

![first_one](https://user-images.githubusercontent.com/17051250/52274973-e8bae700-2956-11e9-8ace-cf60331a0802.png)

in legacy page on the root cms page I can't see any edit action in thumbnails

however when you're on the second page it is visible

![second](https://user-images.githubusercontent.com/17051250/52275232-a04ff900-2957-11e9-90e0-1a77bba736b6.png)

when there is one edit pencil visible I can apply your suggested design however how it should look like if there are more levels to display like in the example below:

![herarchy](https://user-images.githubusercontent.com/17051250/52275459-4bf94900-2958-11e9-8595-e373f6f52872.png)

marionf commented:
@tomas862 

I have an exception on the page

![capture d ecran_977](https://user-images.githubusercontent.com/13449658/52280564-7a792300-295c-11e9-991f-b16c49e767bd.png)

tomas862 committed:
moves position column to another namespace - this prevents from position error appearance

tomas862 commented:
@marionf fixed the red error


@marionf, @TristanLDD  regarding the comments in https://github.com/PrestaShop/PrestaShop/pull/10877#issuecomment-459738302 :

2. Done

regarding the comment https://github.com/PrestaShop/PrestaShop/pull/10877#issuecomment-460600871:

![first_one](https://user-images.githubusercontent.com/17051250/52274973-e8bae700-2956-11e9-8ace-cf60331a0802.png)

in legacy page on the root cms page I can't see any edit action in thumbnails

however when you're on the second page it is visible

![second](https://user-images.githubusercontent.com/17051250/52275232-a04ff900-2957-11e9-90e0-1a77bba736b6.png)

when there is one edit pencil visible I can apply your suggested design however how it should look like if there are more levels to display like in the example below:

![list_thumbainls3](https://user-images.githubusercontent.com/17051250/52328936-02f9d100-29fa-11e9-924d-caf3e983c0dc.png)

marionf commented:
@tomas862 thank you :+1: 

Not sure, if it's this PR or not, we have an alignement issue here:

![capture d ecran_983](https://user-images.githubusercontent.com/13449658/52336052-2d508c00-2a04-11e9-8136-13f22ceff0dd.png)

TristanLDD commented:
Hi @tomas862 

Sorry there is a confusion about the pencil icon because I confused the cms category listing with the category page.. my bad. 

The behavior on this page should actually be the opposite: we should get rid of the pencil icon in the breadcrumb as this component is supposed to be only for navigation, the access to the edit action should be the list view only. Sorry again for the confusion

tomas862 committed:
removes position column usage

tomas862 committed:
removes pencil icons with edit links from breadcrumbs

tomas862 commented:
Hi @marionf . Yes it was a problem with this PR after some updates. Right now I removed the position column usage as this functionality will be improved in another pull request. The list currently looks like this:

![list_breadcrumbs](https://user-images.githubusercontent.com/17051250/52568642-02a56f80-2e18-11e9-97a1-ae4d275e40db.png)

Hi @TristanLDD . I removed pencil icon from breadcrumbs. It is visible in the above picture :)

marionf commented:
@tomas862 

I have this issue on all back-office pages

![capture d ecran_1008](https://user-images.githubusercontent.com/13449658/52578528-ac8af900-2e24-11e9-97d8-0143074d318d.png)

tomas862 committed:
applies php cs fixer

tomas862 commented:
@marionf  strange - not getting this error. Can you check again if the error still exists. I have updated this PR with current develop branch :slightly_smiling_face:

marionf commented:
All good, thank you @tomas862 :+1:

tomas862 commented:
thank you @marionf ! :tada:

matks commented:
Thanks @tomas862 for the hard work on this page 😉
